src/apexguru/apex-guru-service.ts

```diff
@@ -29,22 +33,70 @@ export async function isApexGuruEnabledInOrg(outputChannel: vscode.LogOutputChan
 	}
 }
 
-export async function runApexGuruOnFile(selection: vscode.Uri, outputChannel: vscode.LogOutputChannel) {
+export async function runApexGuruOnFile(selection: vscode.Uri, runInfo: RunInfo) {
+	const {
+		diagnosticCollection,
+		commandName,
+		outputChannel
+	} = runInfo;
+	const startTime = Date.now();
 	try {
-        const requestId = await initiateApexGuruRequest(selection, outputChannel);
-		// TODO: Logging the request Id for easy QA. Future stories will use this requestId to poll and retrieve the Apex Guru report.
-        outputChannel.appendLine('***Apex Guru request Id:***' + requestId);
+		await vscode.window.withProgress({
+			location: vscode.ProgressLocation.Notification
+		}, async (progress) => {
+			progress.report(messages.apexGuru.progress);
+			const connection = await CoreExtensionService.getConnection();
+			const requestId = await initiateApexGuruRequest(selection, outputChannel, connection);
+			outputChannel.appendLine('***Apex Guru request Id:***' + requestId);
+
+			const queryResponse: ApexGuruQueryResponse = await pollAndGetApexGuruResponse(connection, requestId, Constants.APEX_GURU_MAX_TIMEOUT_SECONDS, Constants.APEX_GURU_RETRY_INTERVAL_MILLIS);
+
+			const decodedReport = Buffer.from(queryResponse.report, 'base64').toString('utf8');
+			outputChannel.appendLine('***Retrieved analysis report from ApexGuru***:' + decodedReport);
+
+			const ruleResult = transformStringToRuleResult(selection.fsPath, decodedReport);
+			new DiagnosticManager().displayDiagnostics([selection.fsPath], [ruleResult], diagnosticCollection);
+			TelemetryService.sendCommandEvent(Constants.TELEM_SUCCESSFUL_APEX_GURU_FILE_ANALYSIS, {
+				executedCommand: commandName,
+				duration: (Date.now() - startTime).toString()
+			});
+		});
     } catch (e) {
         const errMsg = e instanceof Error ? e.message : e as string;
         outputChannel.appendLine('***Apex Guru initiate request failed***');
         outputChannel.appendLine(errMsg);
     }
 }
 
-export async function initiateApexGuruRequest(selection: vscode.Uri, outputChannel: vscode.LogOutputChannel): Promise<string> {
+export async function pollAndGetApexGuruResponse(connection: Connection, requestId: string, maxWaitTimeInSeconds: number, retryIntervalInMillis: number): Promise<ApexGuruQueryResponse> {
+	let queryResponse: ApexGuruQueryResponse;
+	let lastErrorMessage = '';
+	const startTime = Date.now();
+	while ((Date.now() - startTime) < maxWaitTimeInSeconds * 1000) {
+		try {
+			queryResponse = await connection.request({
```

Comment:
We should not only follow up with the folks who own this request method regarding the response code... but we should also ask if we can please send in a connection timeout value to ensure that if the server goes down... it isn't going past our remaining timeout time.